File: Samples/Inventory/Unity/Assets/NoesisGUI/Samples/Inventory/DragItemBehavior.cs
#if UNITY_5_3_OR_NEWER
#define NOESIS
using Noesis;
using NoesisApp;
using System.Windows.Input;
#else
using System;
using System.Windows;
using System.Windows.Input;
using Microsoft.Xaml.Behaviors;
#endif

namespace Inventory
{
    class DragItemBehavior : Behavior<FrameworkElement>
    {
        public Point DragStartOffset
        {
            get { return (Point)GetValue(DragStartOffsetProperty); }
            set { SetValue(DragStartOffsetProperty, value); }
        }

        public static readonly DependencyProperty DragStartOffsetProperty = DependencyProperty.Register(
            "DragStartOffset", typeof(Point), typeof(DragItemBehavior),
            new PropertyMetadata(new Point(0, 0)));

        public ICommand StartDragCommand
        {
            get { return (ICommand)GetValue(StartDragCommandProperty); }
            set { SetValue(StartDragCommandProperty, value); }
        }

        public static readonly DependencyProperty StartDragCommandProperty = DependencyProperty.Register(
            "StartDragCommand", typeof(ICommand), typeof(DragItemBehavior),
            new PropertyMetadata(null));

        public ICommand EndDragCommand
        {
            get { return (ICommand)GetValue(EndDragCommandProperty); }
            set { SetValue(EndDragCommandProperty, value); }
        }

        public static readonly DependencyProperty EndDragCommandProperty = DependencyProperty.Register(
            "EndDragCommand", typeof(ICommand), typeof(DragItemBehavior),
            new PropertyMetadata(null));

        protected override void OnAttached()
        {
            base.OnAttached();

            FrameworkElement element = AssociatedObject;
            if (element != null)
            {
                element.GiveFeedback += OnGiveFeedback;
                element.PreviewMouseLeftButtonDown += OnMouseDown;
                element.PreviewMouseLeftButtonUp += OnMouseUp;
                element.PreviewMouseMove += OnMouseMove;
            }
        }

        protected override void OnDetaching()
        {
            FrameworkElement element = AssociatedObject;
            if (element != null)
            {
                element.GiveFeedback -= OnGiveFeedback;
                element.PreviewMouseLeftButtonDown -= OnMouseDown;
                element.PreviewMouseLeftButtonUp -= OnMouseUp;
                element.PreviewMouseMove -= OnMouseMove;
            }

            base.OnDetaching();
        }

        private void OnGiveFeedback(object sender, GiveFeedbackEventArgs e)
        {
            e.UseDefaultCursors = false;
            e.Handled = true;
        }

        private void OnMouseDown(object sender, MouseButtonEventArgs e)
        {
            _mouseClicked = true;
        }

        private void OnMouseUp(object sender, MouseButtonEventArgs e)
        {
            _mouseClicked = false;
        }

        private void OnMouseMove(object sender, MouseEventArgs e)
        {
            if (!_mouseClicked)
            {
                return;
            }

            _mouseClicked = false;

            object item = this.AssociatedObject.DataContext;
            if (item != null)
            {
                if (StartDragCommand != null && StartDragCommand.CanExecute(item))
                {
                    DragStartOffset = e.GetPosition(this.AssociatedObject);

                    StartDragCommand.Execute(item);

#if NOESIS
                    DragDrop.DoDragDrop(this.AssociatedObject, item, DragDropEffects.Move,
                        (source, data, target, dropPoint, effects) =>
                        {
                            bool dragSuccess = effects != DragDropEffects.None;
                            if (EndDragCommand != null && EndDragCommand.CanExecute(dragSuccess))
                            {
                                EndDragCommand.Execute(dragSuccess);
                            }
                        });
#else
                    DragDropEffects effects = DragDrop.DoDragDrop(this.AssociatedObject, item, DragDropEffects.Move);

                    bool dragSuccess = effects != DragDropEffects.None;
                    if (EndDragCommand != null && EndDragCommand.CanExecute(dragSuccess))
                    {
                        EndDragCommand.Execute(dragSuccess);
                    }
#endif
                }
            }
        }

        private bool _mouseClicked = false;
    }
}

File: Samples/Inventory/C++/Src/DragItemBehavior.cpp
////////////////////////////////////////////////////////////////////////////////////////////////////
// NoesisGUI - http://www.noesisengine.com
// Copyright (c) 2013 Noesis Technologies S.L. All Rights Reserved.
////////////////////////////////////////////////////////////////////////////////////////////////////


#include "DragItemBehavior.h"

#include <NsCore/ReflectionImplement.h>
#include <NsCore/TypeId.h>
#include <NsCore/Delegate.h>
#include <NsGui/DragDrop.h>
#include <NsGui/UIElementData.h>
#include <NsGui/PropertyMetadata.h>
#include <NsGui/BaseCommand.h>
#include <NsDrawing/Point.h>


using namespace Inventory;
using namespace Noesis;


////////////////////////////////////////////////////////////////////////////////////////////////////
DragItemBehavior::DragItemBehavior(): _mouseClicked(false)
{
}

////////////////////////////////////////////////////////////////////////////////////////////////////
const Point& DragItemBehavior::GetDragStartOffset() const
{
    return GetValue<Point>(DragStartOffsetProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void DragItemBehavior::SetDragStartOffset(const Point& offset)
{
    SetValue<Point>(DragStartOffsetProperty, offset);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
BaseCommand* DragItemBehavior::GetStartDragCommand() const
{
    return GetValue<Ptr<BaseCommand>>(StartDragCommandProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void DragItemBehavior::SetStartDragCommand(BaseCommand* command)
{
    SetValue<Ptr<BaseCommand>>(StartDragCommandProperty, command);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
BaseCommand* DragItemBehavior::GetEndDragCommand() const
{
    return GetValue<Ptr<BaseCommand>>(EndDragCommandProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void DragItemBehavior::SetEndDragCommand(BaseCommand* command)
{
    SetValue<Ptr<BaseCommand>>(EndDragCommandProperty, command);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
Ptr<Freezable> DragItemBehavior::CreateInstanceCore() const
{
    return *new DragItemBehavior();
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void DragItemBehavior::OnAttached()
{
    ParentClass::OnAttached();

    FrameworkElement* element = GetAssociatedObject();
    element->GiveFeedback() += MakeDelegate(this, &DragItemBehavior::OnGiveFeedback);
    element->PreviewMouseLeftButtonDown() += MakeDelegate(this, &DragItemBehavior::OnMouseDown);
    element->PreviewMouseLeftButtonUp() += MakeDelegate(this, &DragItemBehavior::OnMouseUp);
    element->PreviewMouseMove() += MakeDelegate(this, &DragItemBehavior::OnMouseMove);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void DragItemBehavior::OnDetaching()
{
    FrameworkElement* element = GetAssociatedObject();
    element->GiveFeedback() -= MakeDelegate(this, &DragItemBehavior::OnGiveFeedback);
    element->PreviewMouseLeftButtonDown() -= MakeDelegate(this, &DragItemBehavior::OnMouseDown);
    element->PreviewMouseLeftButtonUp() -= MakeDelegate(this, &DragItemBehavior::OnMouseUp);
    element->PreviewMouseMove() -= MakeDelegate(this, &DragItemBehavior::OnMouseMove);

    ParentClass::OnAttached();
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void DragItemBehavior::OnGiveFeedback(BaseComponent*, const GiveFeedbackEventArgs& e)
{
    e.useDefaultCursors = false;
    e.handled = true;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void DragItemBehavior::OnMouseDown(BaseComponent*, const MouseButtonEventArgs&)
{
    _mouseClicked = true;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void DragItemBehavior::OnMouseUp(BaseComponent*, const MouseButtonEventArgs&)
{
    _mouseClicked = false;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void DragItemBehavior::OnMouseMove(BaseComponent*, const MouseEventArgs& e)
{
    if (!_mouseClicked)
    {
        return;
    }

    _mouseClicked = false;

    FrameworkElement* element = GetAssociatedObject();
    BaseComponent* item = element->GetDataContext();
    if (item != 0)
    {
        BaseCommand* startDrag = GetStartDragCommand();
        if (startDrag != 0 && startDrag->CanExecute(item))
        {
            SetDragStartOffset(element->PointFromScreen(e.position));

            startDrag->Execute(item);

            DragDrop::DoDragDrop(element, item, DragDropEffects_Move, [this](DependencyObject*,
                BaseComponent*, UIElement*, const Point&, uint32_t effects)
            {
                Ptr<BaseComponent> dragSuccess = Boxing::Box<bool>(effects != DragDropEffects_None);
                BaseCommand* endDrag = GetEndDragCommand();
                if (endDrag != 0 && endDrag->CanExecute(dragSuccess))
                {
                    endDrag->Execute(dragSuccess);
                }
            });
        }
    }
}

////////////////////////////////////////////////////////////////////////////////////////////////////
NS_BEGIN_COLD_REGION

NS_IMPLEMENT_REFLECTION(DragItemBehavior)
{
    NsMeta<TypeId>("Inventory.DragItemBehavior");

    UIElementData* data = NsMeta<UIElementData>(TypeOf<SelfClass>());
    data->RegisterProperty<Point>(DragStartOffsetProperty, "DragStartOffset",
        PropertyMetadata::Create(Point(0.0f, 0.0f)));
    data->RegisterProperty<Ptr<BaseCommand>>(StartDragCommandProperty, "StartDragCommand",
        PropertyMetadata::Create(Ptr<BaseCommand>()));
    data->RegisterProperty<Ptr<BaseCommand>>(EndDragCommandProperty, "EndDragCommand",
        PropertyMetadata::Create(Ptr<BaseCommand>()));
}

////////////////////////////////////////////////////////////////////////////////////////////////////
const Noesis::DependencyProperty* DragItemBehavior::DragStartOffsetProperty;
const Noesis::DependencyProperty* DragItemBehavior::StartDragCommandProperty;
const Noesis::DependencyProperty* DragItemBehavior::EndDragCommandProperty;
